File: src/diffusers/models/unets/unet2d/DownEncoderBlock2D.cpp
#include "diffusers/models/unets/unet2d/DownEncoderBlock2D.hpp"
#include "nn/ModuleList.hpp"
#include "nn/SiLU.hpp"
#include "diffusers/models/resnet/ResnetBlock2D.hpp"
#include "diffusers/models/downsampling/Downsample2D.hpp"

DownEncoderBlock2D::DownEncoderBlock2D(
    int64_t in_channels,
    int64_t out_channels,
    float dropout,
    int64_t num_layers,
    float resnet_eps,
    int64_t resnet_groups,
    bool resnet_pre_norm,
    float output_scale_factor,
    bool add_downsample,
    int64_t downsample_padding 
) : add_downsample_(add_downsample)
{
    auto resnets = std::make_shared<ModuleList>(num_layers);
    modules["resnets"] = resnets;

    for (auto i = 0; i < resnets->size(); ++i) {
        auto input_channels = i == 0 ? in_channels : out_channels;

        // if resnet_time_scale_shift == "spatial":
        // else:
        (*resnets)[i] = std::make_shared<ResnetBlock2D<SiLU>>(
            input_channels, // in_channels
            out_channels,
            std::nullopt, // conv_shortcut
            0.0f, // dropout
            std::nullopt, // temb_channels
            resnet_groups, // groups
            std::nullopt, // groups_out
            resnet_pre_norm, // per_norm
            resnet_eps, // eps
            false, // skip_time_act
            3, // kernel_size
            output_scale_factor,
            std::nullopt, // use_in_shortcut
            false, // up
            false, // down
            true, // conv_shortcut_bias
            std::nullopt // conv_2d_out_channels
        );
    }
    
    if (add_downsample)
        modules["downsamplers"] = std::shared_ptr<Module>(new ModuleList({
            std::shared_ptr<Module>(new Downsample2D(
                out_channels, // channels
                true, // use_conv
                out_channels,
                downsample_padding // padding
            ))
        }));
}

Tensor DownEncoderBlock2D::forward(Context& context, Tensor hidden_states) {
    auto resnets = std::static_pointer_cast<ModuleList>(modules["resnets"]);

    for (auto i = 0; i < resnets->size(); ++i)
        hidden_states = std::static_pointer_cast<ResnetBlock2D<SiLU>>((*resnets)[i])->forward(context, hidden_states);

    if (add_downsample_) {
        auto downsamplers = std::static_pointer_cast<ModuleList>(modules["downsamplers"]);

        for (auto i = 0; i < downsamplers->size(); ++i)
            hidden_states = std::static_pointer_cast<Downsample2D>((*downsamplers)[i])->forward(context, hidden_states);
    }

    return hidden_states;
}

File: test/src/diffusers/models/unets/unet2d/DownEncoderBlock2D.py
from utils import TestCase
import torch
from diffusers.models.unets.unet_2d_blocks import DownEncoderBlock2D

class TestNNDownEncoderBlock2D(TestCase):
    def test_basic(self):
        model = DownEncoderBlock2D(
            num_layers=2,
            in_channels=4,
            out_channels=8,
            add_downsample=False,
            resnet_eps=1e-6,
            resnet_act_fn="silu",
            resnet_groups=4,
        )

        hidden_states = torch.randn(1, 4, 8, 8)

        expected = model.forward(hidden_states)

        actual = self.cli(
            "DownEncoderBlock2D",
            "--num_layers", "2",
            "--in_channels", "4",
            "--out_channels", "8",
            "--add_downsample", "false",
            "--resnet_groups", "4",
            "--hidden_states", str(hidden_states.tolist()),
            *self.params(model)
        )

        self.assertTensors(actual, [expected])

    def test_with_downsample(self):
        model = DownEncoderBlock2D(
            num_layers=2,
            in_channels=4,
            out_channels=8,
            add_downsample=True,
            resnet_eps=1e-6,
            resnet_act_fn="silu",
            resnet_groups=4,
        )

        hidden_states = torch.randn(1, 4, 16, 16)

        expected = model.forward(hidden_states)

        actual = self.cli(
            "DownEncoderBlock2D",
            "--num_layers", "2",
            "--in_channels", "4",
            "--out_channels", "8",
            "--add_downsample", "true",
            "--resnet_groups", "4",
            "--hidden_states", str(hidden_states.tolist()),
            *self.params(model)
        )

        self.assertTensors(actual, [expected])

    def test_with_downsample_padding_zero(self):
        model = DownEncoderBlock2D(
            num_layers=2,
            in_channels=8,
            out_channels=8,
            add_downsample=True,
            downsample_padding=0,
            resnet_eps=1e-6,
            resnet_act_fn="silu",
            resnet_groups=4,
        )

        hidden_states = torch.randn(1, 8, 16, 16)

        expected = model.forward(hidden_states)

        actual = self.cli(
            "DownEncoderBlock2D",
            "--num_layers", "2",
            "--in_channels", "8",
            "--out_channels", "8",
            "--add_downsample", "true",
            "--downsample_padding", "0",
            "--resnet_groups", "4",
            "--hidden_states", str(hidden_states.tolist()),
            *self.params(model)
        )

        self.assertTensors(actual, [expected])